runtime/runtime.go

```diff
@@ -83,165 +58,58 @@ func getFunctionIO(ctx context.Context) (*Runtime, error) {
-	return obj.AddToScheme(s)
-}
-
-// GetFromObservedKubeObject gets the k8s resource o from a provider kubernetes object kon (Kube Object Name)
-// from the observed array of the FunctionIO.
-func (in *Runtime) GetFromObservedKubeObject(ctx context.Context, o client.Object, kon string) error {
-	ko, err := in.getKubeObjectFrom(ctx, in.observedRessources(), o, kon)
-	if err != nil {
-		return err
-	}
-	return in.fromKubeObjectObserved(ko, o)
-}
-
-// GetFromDesiredKubeObject gets the k8s resource o from a provider kubernetes object kon (Kube Object Name)
-// from the desired array of the FunctionIO.
-func (in *Runtime) GetFromDesiredKubeObject(ctx context.Context, o client.Object, kon string) error {
-	ko, err := in.getKubeObjectFrom(ctx, in.desiredRessources(), o, kon)
-	if err != nil {
-		return err
-	}
-	return in.fromKubeObjectDesired(ko, o)
-}
-
-func (in *Runtime) getKubeObjectFrom(ctx context.Context, ressources []funcRessource, o client.Object, kon string) (*xkube.Object, error) {
-	log := controllerruntime.LoggerFrom(ctx)
-
-	log.V(1).Info("Creating kube object from name and unmarshalling it", "kube object", kon)
-	ko := &xkube.Object{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       xkube.ObjectKind,
-			APIVersion: xkube.ObjectKindAPIVersion,
-		},
-	}
-	err := in.getFrom(ressources, ko, kon)
-	if err != nil {
-		return nil, fmt.Errorf("cannot get unmarshall kubernetes object: %w", err)
+	r.Observed = ObservedResources{
+		resources: *observedResources(r.io.Observed.Resources),
+		composite: r.io.Observed.Composite,
 	}
-
-	log.V(1).Info("Unmarshalling object from kube object", "object type", reflect.TypeOf(o))
-	return ko, nil
-}
-
-func (in *Runtime) desiredRessources() []funcRessource {
-	ressources := make([]funcRessource, len(in.Desired.Resources))
-
-	for i := range in.Desired.Resources {
-		ressources[i] = &desiredRessource{DesiredResource: in.Desired.Resources[i]}
+	r.Desired = DesiredResources{
+		resources: *desiredResources(r.io.Desired.Resources),
+		composite: r.io.Desired.Composite,
 	}
 
-	return ressources
+	return &r, nil
 }
 
-func (in *Runtime) observedRessources() []funcRessource {
-	ressources := make([]funcRessource, len(in.Observed.Resources))
-
-	for i := range in.Observed.Resources {
-		ressources[i] = &observedResource{ObservedResource: in.Observed.Resources[i]}
-	}
-
-	return ressources
-}
-
-func (in *Runtime) fromKubeObjectObserved(kobj *xkube.Object, obj client.Object) error {
+func fromKubeObject(ctx context.Context, kobj *xkube.Object, obj client.Object) error {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.V(1).Info("Unmarshalling resource from kube object", "kube object", kobj, reflect.TypeOf(obj).Kind())
 	if kobj.Status.AtProvider.Manifest.Raw == nil {
-		return fmt.Errorf("no resource in kubernetes object")
+		if kobj.Spec.ForProvider.Manifest.Raw == nil {
+			return fmt.Errorf("no resource in kubernetes object")
+		}
+		return json.Unmarshal(kobj.Spec.ForProvider.Manifest.Raw, obj)
 	}
```

Comment:
This is too much magic for me. When I get a kube.Object, I have no Idea if I get the spec or status manifest. 

IMO when I get a kube object from `Desired` I should get the spec and if I get it from `Observed` I should get the status.

Reply:
I agree this is not the best approach here to query both status and spec but I would just get the status instead of spec. We need to make sure we get what the actual resource will be when it's persisted.

Reply:
I don't think I agree with that. If I get the `Desired` I'm interested in the `spec`. If we always return the `status`, I'm not sure why we'd allow to get kube.Objects from `Desired` anyways as it's always the same as in `Observered`.

Let's say the base composition adds a kube.Object containing with a config map. Now the composition function wants to add another field based on the existing fields, which isn't possible to construct using patches. Now I get the `Desired` resource, but on the first iteration it will be empty as there is no `status` so I can't add the field, but have to wait until the reconcile loop runs again. In this case I explicitly want to get the desired object, so IMO I should get the spec.

Reply:
> but on the first iteration it will be empty as there is no `status`

Cannot argue against this. So you propose to add get the spec from Desired but status from Observed?

Reply:
Yes. Which IMO is intuitive because the `spec` is the desired state of the kube.Object and the `status` is the observed state